plugin/rewrite/name.go

```diff
@@ -58,62 +114,62 @@ const (
 
 // Rewrite rewrites the current request based upon exact match of the name
 // in the question section of the request.
-func (rule *exactNameRule) Rewrite(ctx context.Context, state request.Request) Result {
+func (rule *exactNameRule) Rewrite(ctx context.Context, state request.Request) (ResponseRule, Result) {
 	if rule.From == state.Name() {
 		state.Req.Question[0].Name = rule.To
-		return RewriteDone
+		return rule.Response(state)
 	}
-	return RewriteIgnored
+	return nil, RewriteIgnored
 }
 
 // Rewrite rewrites the current request when the name begins with the matching string.
-func (rule *prefixNameRule) Rewrite(ctx context.Context, state request.Request) Result {
+func (rule *prefixNameRule) Rewrite(ctx context.Context, state request.Request) (ResponseRule, Result) {
 	if strings.HasPrefix(state.Name(), rule.Prefix) {
 		state.Req.Question[0].Name = rule.Replacement + strings.TrimPrefix(state.Name(), rule.Prefix)
-		return RewriteDone
+		return rule.Response(state)
 	}
-	return RewriteIgnored
+	return nil, RewriteIgnored
 }
 
 // Rewrite rewrites the current request when the name ends with the matching string.
-func (rule *suffixNameRule) Rewrite(ctx context.Context, state request.Request) Result {
+func (rule *suffixNameRule) Rewrite(ctx context.Context, state request.Request) (ResponseRule, Result) {
 	if strings.HasSuffix(state.Name(), rule.Suffix) {
 		state.Req.Question[0].Name = strings.TrimSuffix(state.Name(), rule.Suffix) + rule.Replacement
-		return RewriteDone
+		return rule.Response(state)
 	}
-	return RewriteIgnored
+	return nil, RewriteIgnored
 }
 
 // Rewrite rewrites the current request based upon partial match of the
 // name in the question section of the request.
-func (rule *substringNameRule) Rewrite(ctx context.Context, state request.Request) Result {
+func (rule *substringNameRule) Rewrite(ctx context.Context, state request.Request) (ResponseRule, Result) {
 	if strings.Contains(state.Name(), rule.Substring) {
 		state.Req.Question[0].Name = strings.Replace(state.Name(), rule.Substring, rule.Replacement, -1)
-		return RewriteDone
+		return rule.Response(state)
 	}
-	return RewriteIgnored
+	return nil, RewriteIgnored
 }
 
 // Rewrite rewrites the current request when the name in the question
 // section of the request matches a regular expression.
-func (rule *regexNameRule) Rewrite(ctx context.Context, state request.Request) Result {
+func (rule *regexNameRule) Rewrite(ctx context.Context, state request.Request) (ResponseRule, Result) {
 	regexGroups := rule.Pattern.FindStringSubmatch(state.Name())
 	if len(regexGroups) == 0 {
-		return RewriteIgnored
+		return nil, RewriteIgnored
 	}
 	s := rule.Replacement
 	for groupIndex, groupValue := range regexGroups {
 		groupIndexStr := "{" + strconv.Itoa(groupIndex) + "}"
 		s = strings.Replace(s, groupIndexStr, groupValue, -1)
 	}
 	state.Req.Question[0].Name = s
-	return RewriteDone
+	return rule.Response(state)
```

Comment:
why does this need to be returned? `rule.Response(state)` can also be called outside of this method

Reply:
This is a dedicated feature for the name rewrite types, that need to dynamically handle the auto and non-auto case. I've separated this into this response function, because it is identical for all name types. Response is just an attribute. Basically all the attributes should be internal(lower case), but they were public so so far, therefore I didn't changed this.

Reply:
May be it is better to encapsulate the common behaviour of those name rules into a dedicated base struct used for the name rule types. I could provide this, It could also handle the `Mode()` function.

Reply:
I've changed this now and introduced a common base part, which allowed to get rid of even more coding

Reply:
making members/functions private is OK w/ me. I think this was mostly a side effect of style. There should be nothing consuming this as an API